module/applications/components/damage-application.mjs

```diff
@@ -0,0 +1,334 @@
+
+    // Build the frame HTML only once
+    if ( !this.targetList ) {
+      const div = document.createElement("div");
+      div.classList.add("card-tray", "damage-tray", "collapsible");
+      div.innerHTML = `
+        <label class="roboto-upper">
+          <i class="fa-solid fa-heart-crack"></i>
+          <span>${game.i18n.localize("DND5E.Apply")}</span>
+          <i class="fa-solid fa-caret-down"></i>
+        </label>
+        <div class="collapsible-content">
+          <div class="target-source-control">
+            <button type="button" class="unbutton" data-mode="targeted" aria-pressed="true">
+              <i class="fa-solid fa-bullseye" inert></i> ${game.i18n.localize("DND5E.Tokens.Targeted")}
+            </button>
+            <button type="button" class="unbutton" data-mode="selected" aria-pressed="false">
+              <i class="fa-solid fa-users-viewfinder inert"></i> ${game.i18n.localize("DND5E.Tokens.Selected")}
+            </button>
+          </div>
+          <ul class="targets unlist"></ul>
+          <button class="apply-damage" type="button" data-action="applyDamage">
+            <i class="fa-solid fa-reply-all fa-flip-horizontal inert"></i>
+            ${game.i18n.localize("DND5E.Apply")}
+          </button>
+        </div>
+      `;
+      this.replaceChildren(div);
+      this.applyButton = div.querySelector(".apply-damage");
+      this.applyButton.addEventListener("click", this._onApplyDamage.bind(this));
+      this.targetList = div.querySelector(".targets");
+      this.targetSourceControl = this.querySelector(".target-source-control");
+      this.targetSourceControl.querySelectorAll("button").forEach(b =>
+        b.addEventListener("click", this._onChangeTargetMode.bind(this))
+      );
+      if ( !game.user.isGM || !this.chatMessage.getFlag("dnd5e", "targets")?.length ) {
+        this.targetSourceControl.hidden = true;
+      }
+      div.querySelector(".collapsible-content").addEventListener("click", event => {
+        event.stopImmediatePropagation();
+      });
+    }
+
+    this.buildTargetsList();
+  }
+
+  /* -------------------------------------------- */
+
+  /**
+   * Build a list of targeted tokens based on current mode & replace any existing targets.
+   */
+  async buildTargetsList() {
+    // Ensure collapsible section automatically resizes when content changes
+    const collapsible = this.querySelector(".collapsible-content");
+    if ( collapsible.style.height !== "0px" ) collapsible.style.height = "auto";
+
+    let targetedTokens;
+    switch ( this.targetingMode ) {
+      case "targeted":
+        targetedTokens = (this.chatMessage.getFlag("dnd5e", "targets") ?? []).map(t => t.uuid);
+        if ( this.selectedTokensHook ) {
+          Hooks.off("controlToken", this.selectedTokensHook);
+          this.selectedTokensHook = null;
+        }
+        break;
+      case "selected":
+        targetedTokens = canvas.tokens.controlled.map(t => t.document.actor.uuid);
```

Comment:
Did some quick fooling around, and this lead to some weirdness with multiple tokens of the same actor being duplicated (as well as tokens with no actor throwing errors).

Reply:
When you say 'multiple tokens of the same actor', I assume you mean multiple *linked* tokens of the same actor? Unlinked token actors should have unique UUIDs from each other. We should de-dupe those though since applying damage to any one of a linked actor's tokens applies damage to all of them. And we should fix the error for actor-less tokens of course.

Reply:
Yeah sorry, that is what I meant. I fixed an identical issue [here](https://github.com/foundryvtt/dnd5e/pull/3090).